test/hooks.js

```diff
@@ -0,0 +1,682 @@
+'use strict'
+
+const { test } = require('tap')
+const Fastify = require('fastify')
+const { parse, GraphQLSchema } = require('graphql')
+const GQL = require('..')
+
+const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
+
+process.removeAllListeners('warning')
+
+const schema = `
+  type Query {
+    add(x: Int, y: Int): Int
+  }
+`
+
+const resolvers = {
+  Query: {
+    add: async (_, obj) => {
+      const { x, y } = obj
+      return x + y
+    }
+  }
+}
+
+const query = '{ add(x: 2, y: 2) }'
+
+async function createTestServer (t, customResolvers = resolvers) {
+  const app = Fastify()
+  t.tearDown(() => {
+    app.close()
+  })
+
+  app.register(GQL, { schema, resolvers: customResolvers })
+
+  // needed so that graphql is defined
+  await app.ready()
+
+  return app
+}
+
+// -----
+// hooks
+// -----
+test('hooks', async t => {
+  t.plan(16)
+  const app = await createTestServer(t)
+
+  app.graphql.addHook('preParsing', async function (schema, source, context) {
+    await sleep(1)
```

Comment:
why is this needed?

Reply:
I wanted to have an explicit promise to await within the hooks function so we can simulate some work - similar to how fastify tests hooks. I'll remove it however, because it's not really needed for these test cases and what is currently there in the function definitions should be sufficient!

Reply:
You can do:

```js
const immediate = promisify(setImmediate).
```

Reply:
ah nice - thanks! I'll push up the rest of these changes you've suggested now and get to work on adding thorough tests for subscriptions